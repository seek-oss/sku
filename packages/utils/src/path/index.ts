export { toPosixPath } from './toPosixPath.js';
