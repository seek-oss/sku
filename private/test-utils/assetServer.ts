import handler from 'serve-handler';
import { createServer } from 'node:http';

// Not exported from `serve-handler` for some reason
type Rewrite = {
  source: string;
  destination: string;
};

export const startAssetServer = async (
  port: number,
  targetDirectory: string,
  rewrites: Rewrite[] = [],
) => {
  const server = createServer((request, response) =>
    handler(request, response, {
      public: targetDirectory,
      // So we can test storybook iframe pages when serving a built storybook
      cleanUrls: ['!/iframe.html'],
      rewrites,
      headers: [
        {
          source: '**/*.*',
          headers: [
            {
              key: 'Access-Control-Allow-Origin',
              value: '*',
            },
          ],
        },
      ],
    }),
  );

  server.listen(port);

  return () => {
    server.close((e) => {
      if (e) {
        console.error('Error closing asset server', e);
      }
    });
  };
};
