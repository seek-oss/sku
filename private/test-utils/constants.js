/**
 * This file is used to store constants that are shared across test configs and test files.
 * This needs to be a JS file so that it can be imported in the vitest config.
 */

export const TEST_TIMEOUT = 30_000;
