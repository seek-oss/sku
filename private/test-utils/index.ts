export { ListExternalsWebpackPlugin } from './ListExternalsWebpackPlugin.ts';
export { dirContentsToObject } from './dirContentsToObject.ts';
export { makeStableHashes, makeStableViteHashes } from './skuConfig.ts';
export { getStoryPage, getTextContentFromFrameOrPage } from './storybook.ts';
export { getPort } from './getPort.ts';
