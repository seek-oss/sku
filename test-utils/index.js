// @ts-check
export { ListExternalsWebpackPlugin } from './ListExternalsWebpackPlugin.js';
export { getAppSnapshot } from './appSnapshot.js';
export { startAssetServer } from './assetServer.js';
export { dirContentsToObject } from './dirContentsToObject.js';
export { run, runSkuScriptInDir } from './process.ts';
export { makeStableHashes } from './skuConfig.js';
export { getStoryPage, getTextContentFromFrameOrPage } from './storybook.js';
export { waitForUrls } from './waitForUrls.js';
