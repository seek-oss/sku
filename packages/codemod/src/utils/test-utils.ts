import { readFile } from 'node:fs/promises';
import path from 'node:path';
import { expect } from 'vitest';
import { isValidTransform, type Transform } from './types.js';

type FileInfo = {
  source: string;
  path: string;
};

type InlineTestProps = {
  transform: Transform;
  input: FileInfo;
  expectedOutput: string;
};

type TestProps = {
  dirName: string;
  transformName: string;
  fixtureName?: string;
  testOptions?: { extension?: string };
};

const runInlineTest = async ({
  transform,
  input,
  expectedOutput,
}: InlineTestProps) => {
  if (!isValidTransform(transform)) {
    throw new Error('Invalid transform function provided');
  }

  const output = transform(input.source);
  if (!output) {
    throw new Error('Transform failed');
  }

  expect(output.trim()).toEqual(expectedOutput.trim());

  return output;
};

export const runTest = async ({
  dirName,
  transformName,
  fixtureName,
  testOptions = {},
}: TestProps) => {
  const testFixtureName = fixtureName || transformName;

  const fixtureDir = path.join(dirName, '..', '__testfixtures__');
  const inputPath = path.join(
    fixtureDir,
    `${testFixtureName}.input.${testOptions.extension}`,
  );

  // Assumes transform is one level up from __tests__ directory
  const transformImport = import(path.join(dirName, '..', transformName));
  const sourceRead = readFile(inputPath, 'utf8');
  const expectedOutputRead = readFile(
    path.join(fixtureDir, `${testFixtureName}.output.${testOptions.extension}`),
    'utf8',
  );

  const [{ transform }, source, expectedOutput] = await Promise.all([
    transformImport,
    sourceRead,
    expectedOutputRead,
  ]);

  return runInlineTest({
    transform,
    input: {
      path: inputPath,
      source,
    },
    expectedOutput,
  });
};

export const runNoChangeTest = async ({
  dirName,
  transformName,
  fixtureName,
  testOptions = {},
}: TestProps) => {
  const inputPath = path.join(
    dirName,
    '..',
    '__testfixtures__',
    `${fixtureName}.unchanged.${testOptions.extension}`,
  );

  const sourceRead = readFile(inputPath, 'utf8');
  const transformImport = import(path.join(dirName, '..', transformName));

  const [{ transform }, source] = await Promise.all([
    transformImport,
    sourceRead,
  ]);
  const output = await transform(source);

  expect(output).toEqual(null);
};
