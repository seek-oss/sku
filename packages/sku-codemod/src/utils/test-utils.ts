import { readFile } from 'node:fs/promises';
import path from 'node:path';
import { expect } from 'vitest';

type FileInfo = {
  source: string;
  path: string;
};

const runInlineTest = async (
  transform: (source: string) => string | false,
  input: FileInfo,
  expectedOutput: string,
) => {
  const output = await transform(input.source);
  if (!output) {
    throw new Error('Transform failed');
  }
  const expectation = (o: string) =>
    expect(o.trim()).toEqual(expectedOutput.trim());
  expectation(output);
  return output;
};

export const runTest = async (
  dirName: string,
  transformName: string,
  testFilePrefix?: string,
  testOptions: Record<string, unknown> = {},
) => {
  let localTestFilePrefix = testFilePrefix;
  if (!testFilePrefix) {
    localTestFilePrefix = transformName;
  }

  // Assumes transform is one level up from __tests__ directory
  const transform = (await import(path.join(dirName, '..', transformName)))
    .transform;
  const fixtureDir = path.join(dirName, '..', '__testfixtures__');
  const inputPath = path.join(
    fixtureDir,
    `${localTestFilePrefix}.input.${testOptions.extension}`,
  );
  const source = await readFile(inputPath, 'utf8');
  const expectedOutput = await readFile(
    path.join(
      fixtureDir,
      `${localTestFilePrefix}.output.${testOptions.extension}`,
    ),
    'utf8',
  );
  return runInlineTest(
    transform,
    {
      path: inputPath,
      source,
    },
    expectedOutput,
  );
};

export const runNoChangeTest = async (
  dirName: string,
  transformName: string,
  testFilePrefix: string,
  testOptions: Record<string, unknown>,
) => {
  const transform = (await import(path.join(dirName, '..', transformName)))
    .transform;
  const inputPath = path.join(
    dirName,
    '..',
    '__testfixtures__',
    `${testFilePrefix}.unchanged.${testOptions.extension}`,
  );
  const source = await readFile(inputPath, 'utf8');
  const output = await transform(source);
  expect(output).toEqual(false);
};
