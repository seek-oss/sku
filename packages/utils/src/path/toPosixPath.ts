import path from 'node:path';

/**
 * Replaces all win32 path separators (\) with posix path separators (/)
 */
export const toPosixPath = (inputPath: string) =>
  inputPath.replaceAll(path.win32.sep, path.posix.sep);
