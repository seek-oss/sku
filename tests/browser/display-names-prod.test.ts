import {
  describe,
  beforeAll,
  afterAll,
  it,
  expect as globalExpected,
} from 'vitest';
import { readFileSync, readdirSync } from 'node:fs';
import { join } from 'node:path';
import { dirContentsToObject } from '@sku-private/test-utils';
import {
  scopeToFixture,
  bundlers,
  cleanup,
  type BundlerValues,
} from '@sku-private/testing-library';

const { sku, fixturePath } = scopeToFixture('display-names-prod');

describe('display-names-prod', () => {
  describe.sequential.for(bundlers)('bundler: %s', (bundler) => {
    const args: BundlerValues<string[]> = {
      vite: ['--config', 'sku.config.vite.ts', '--experimental-bundler'],
      webpack: [],
    };

    describe('build', () => {
      beforeAll(async () => {
        const build = await sku('build', args[bundler]);
        globalExpected(
          await build.findByText('Sku build complete'),
        ).toBeInTheConsole();
      });

      it('should create build output', async ({ expect }) => {
        const distDir = fixturePath('dist');
        const files = await dirContentsToObject(distDir);
        expect(files).toMatchSnapshot();
      });

      it('should add displayName assignments to built JavaScript', async ({
        expect,
      }) => {
        const distDir = fixturePath('dist');
        const allFiles = readdirSync(distDir);

        const jsFiles = allFiles
          .filter((filename) => filename.endsWith('.js'))
          .map((filename) => ({
            filename,
            content: readFileSync(join(distDir, filename), 'utf-8'),
          }));

        expect(jsFiles.length).toBeGreaterThan(0);

        const hasDisplayNameAssignments = jsFiles.some(({ content }) =>
          /\.displayName\s*=\s*['"`][^'"`]+['"`]/.test(content),
        );

        expect(hasDisplayNameAssignments).toBe(true);

        const allJsContent = jsFiles.map(({ content }) => content).join('\n');
        expect(allJsContent).toMatch(/\w+\.displayName\s*=\s*['"`]Button['"`]/);
        expect(allJsContent).toMatch(/\w+\.displayName\s*=\s*['"`]Card['"`]/);
        expect(allJsContent).toMatch(/\w+\.displayName\s*=\s*['"`]Header['"`]/);
      });
    });
  });

  afterAll(cleanup);
});
