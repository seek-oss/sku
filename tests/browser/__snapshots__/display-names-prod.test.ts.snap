// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`display-names-prod > bundler: vite > build > should add displayName properties to React components in production build 1`] = `
"<!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8">
    <title>
      hello-world
    </title>
    <meta
      name="viewport"
      content="width=device-width, initial-scale=1"
    >
    <link
      rel="modulepreload"
      href="/vite-client-ljKe-wB2.js"
      crossorigin
    >
  </head>
  <body>
    <div id="app">
      <div>
        <h1>
          Display Names Test
        </h1>
        <button>
          Click me
        </button>
        <div>
          This is a card
        </div>
        <input placeholder="Test input">
        <footer>
          Footer content
        </footer>
        <div id="display-names-test">
          <p>
            Components should have displayName in production build:
          </p>
          <ul>
            <li>
              Button.displayName:
              Button
            </li>
            <li>
              Card.displayName:
              Card
            </li>
            <li>
              Header.displayName:
              Header
            </li>
            <li>
              MemoizedFooter.displayName:
              MemoizedFooter
            </li>
            <li>
              Input.displayName:
              Input
            </li>
          </ul>
        </div>
      </div>
    </div>
    <script
      type="module"
      src="/vite-client-ljKe-wB2.js"
    >
    </script>
  </body>
</html>"
POST HYDRATE DIFFS: NO DIFF
`;

exports[`display-names-prod > bundler: vite > build > should create build output with displayNames 1`] = `
{
  "index.html": "<!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8" />
    <title>hello-world</title>
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <link rel="modulepreload" href="/vite-client-ljKe-wB2.js" crossorigin />

  </head>
  <body>
    <div id="app"><div><h1>Display Names Test</h1><button>Click me</button><div>This is a card</div><input placeholder="Test input"/><footer>Footer content</footer><div id="display-names-test"><p>Components should have displayName in production build:</p><ul><li>Button.displayName: <!-- -->Button</li><li>Card.displayName: <!-- -->Card</li><li>Header.displayName: <!-- -->Header</li><li>MemoizedFooter.displayName:<!-- --> <!-- -->MemoizedFooter</li><li>Input.displayName: <!-- -->Input</li></ul></div></div></div>
    <script type="module" src="/vite-client-ljKe-wB2.js" ></script>
  </body>
</html>",
  "vite-client-ljKe-wB2.js": "CONTENTS IGNORED IN SNAPSHOT TEST",
  "vite-client-ljKe-wB2.js.map": "CONTENTS IGNORED IN SNAPSHOT TEST",
}
`;

exports[`display-names-prod > bundler: webpack > build > should add displayName properties to React components in production build 1`] = `
"<!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8">
    <title>
      hello-world
    </title>
    <meta
      name="viewport"
      content="width=device-width, initial-scale=1"
    >
    <link
      data-chunk="main"
      rel="preload"
      as="script"
      href="/runtime-980fb1ffe39d4de310ac.js"
    >
    <link
      data-chunk="main"
      rel="preload"
      as="script"
      href="/7-4274d242a3d5e5313473.js"
    >
    <link
      data-chunk="main"
      rel="preload"
      as="script"
      href="/main-6071e23940ef35af1f55.js"
    >
  </head>
  <body>
    <div id="app">
      <div>
        <h1>
          Display Names Test
        </h1>
        <button>
          Click me
        </button>
        <div>
          This is a card
        </div>
        <input placeholder="Test input">
        <footer>
          Footer content
        </footer>
        <div id="display-names-test">
          <p>
            Components should have displayName in production build:
          </p>
          <ul>
            <li>
              Button.displayName:
              Button
            </li>
            <li>
              Card.displayName:
              Card
            </li>
            <li>
              Header.displayName:
              Header
            </li>
            <li>
              MemoizedFooter.displayName:
              MemoizedFooter
            </li>
            <li>
              Input.displayName:
              Input
            </li>
          </ul>
        </div>
      </div>
    </div>
    <script
      id="__LOADABLE_REQUIRED_CHUNKS__"
      type="application/json"
    >
      []
    </script>
    <script
      id="__LOADABLE_REQUIRED_CHUNKS___ext"
      type="application/json"
    >
      {"namedChunks":[]}
    </script>
    <script
      async
      data-chunk="main"
      src="/runtime-980fb1ffe39d4de310ac.js"
    >
    </script>
    <script
      async
      data-chunk="main"
      src="/7-4274d242a3d5e5313473.js"
    >
    </script>
    <script
      async
      data-chunk="main"
      src="/main-6071e23940ef35af1f55.js"
    >
    </script>
  </body>
</html>"
POST HYDRATE DIFFS: NO DIFF
`;

exports[`display-names-prod > bundler: webpack > build > should create build output with displayNames 1`] = `
{
  "7-4274d242a3d5e5313473.js": "CONTENTS IGNORED IN SNAPSHOT TEST",
  "7-4274d242a3d5e5313473.js.map": "CONTENTS IGNORED IN SNAPSHOT TEST",
  "index.html": "<!DOCTYPE html>
<html>
  <head>
    <meta charset="UTF-8" />
    <title>hello-world</title>
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    
<link data-chunk="main" rel="preload" as="script" href="/runtime-980fb1ffe39d4de310ac.js">
<link data-chunk="main" rel="preload" as="script" href="/7-4274d242a3d5e5313473.js">
<link data-chunk="main" rel="preload" as="script" href="/main-6071e23940ef35af1f55.js">
  </head>
  <body>
    <div id="app"><div><h1>Display Names Test</h1><button>Click me</button><div>This is a card</div><input placeholder="Test input"/><footer>Footer content</footer><div id="display-names-test"><p>Components should have displayName in production build:</p><ul><li>Button.displayName: <!-- -->Button</li><li>Card.displayName: <!-- -->Card</li><li>Header.displayName: <!-- -->Header</li><li>MemoizedFooter.displayName:<!-- --> <!-- -->MemoizedFooter</li><li>Input.displayName: <!-- -->Input</li></ul></div></div></div>
    <script id="__LOADABLE_REQUIRED_CHUNKS__" type="application/json">[]</script><script id="__LOADABLE_REQUIRED_CHUNKS___ext" type="application/json">{"namedChunks":[]}</script>
<script async data-chunk="main" src="/runtime-980fb1ffe39d4de310ac.js"></script>
<script async data-chunk="main" src="/7-4274d242a3d5e5313473.js"></script>
<script async data-chunk="main" src="/main-6071e23940ef35af1f55.js"></script>
  </body>
</html>",
  "main-6071e23940ef35af1f55.js": "CONTENTS IGNORED IN SNAPSHOT TEST",
  "main-6071e23940ef35af1f55.js.map": "CONTENTS IGNORED IN SNAPSHOT TEST",
  "runtime-980fb1ffe39d4de310ac.js": "CONTENTS IGNORED IN SNAPSHOT TEST",
  "runtime-980fb1ffe39d4de310ac.js.map": "CONTENTS IGNORED IN SNAPSHOT TEST",
}
`;
